### ISSUE TITLE
Floats in multitask throwing SIMD FP exception even after adding `fxsave` and `fxrstor` to `mt_switch`

### ISSUE THREAD

**Issue by 8dcc** [fs-os/fs-os#14]
Running the `test_multitask` command on the latest commit (15bb14709b1655d4b2be047600f026604ee29e9e) seems to work fine, because it's only printing integers:

![integers](https://github.com/fs-os/fs-os/assets/29655971/bd4f4a4f-f87b-4c0c-afbf-e7abad7508cb)

If we print floats instead, the following exception occurs:

![exception](https://github.com/fs-os/fs-os/assets/29655971/ef061ff0-04c8-467a-bd1c-166403ed907a)

If I check what that address is with my beloved [rizin](https://github.com/rizinorg/rizin) debugger/decompiler, we can see the following when decompiling 10 instructions at `0x1044DB`:

![rizin](https://github.com/fs-os/fs-os/assets/29655971/b402449a-3e60-43a2-b09b-018f5325eab5)

That is the disassembly of the `fmt_f` function, used for printing the `"%f"` formats by `vprintf()`:

![fmt_f](https://github.com/fs-os/fs-os/assets/29655971/fc88ba54-6e60-4746-b65f-eb5106a8d95e)

The instruction causing the exception is `cvttsd2si`. From [felixcloutier.com](https://www.felixcloutier.com/x86/cvttsd2si):

> Converts a double precision floating-point value in the source operand (the second operand) to a signed double-word integer (or signed quadword integer if operand size is 64 bits) in the destination operand (the first operand).
>
> The source operand can be an XMM register or a 64-bit memory location. The destination operand is a general purpose register.

This assembly was generated by the [`int int_part = (int)num;`](https://github.com/fs-os/fs-os/blob/main/src/libk/stdio.c#L118) expression on top of `fmt_f`. Now I know where the exception is happening, and that it is obviously related to SSE in some way.

It's also important to note that this only happens when trying to print a float after using creating and switching tasks. You can print floats from `kernel_main` without issue.

From the intel manual (*which is an amazing resource that I always forget exists*), at **Chapter 11.5.1** of the volume 1, we can read the following:

> SIMD floating-point exceptions are those exceptions that can be generated by SSE/SSE2/SSE3 instructions that operate on packed or scalar floating-point operands.
> Six classes of SIMD floating-point exceptions can be generated:
>   - Invalid operation
>   - Divide-by-zero
>   - Denormal operand
>   - Numeric overflow
>   - Numeric underflow
>   - Inexact result (Precision)
>
> All SIMD floating-point exceptions are precise and occur as soon as the instruction completes execution.
> Each of the six exception conditions has a corresponding flag [...] in the MXCSR register (see Figure 10-3).
>
> ![Figure 10-3](https://github.com/fs-os/fs-os/assets/29655971/82b72add-a5b6-403f-91fe-9e596646d429)

There is more information on the `MXCSR` register. If we patch the exception wrapper a bit, we can use `stmxcsr [dword_addr]` to print the contents of `MXCSR`:

```
0xB94401
0b00000000101110010100010000000001
                            ^^^^^^
```

Those last 6 bits should indicate the type of SIMD floating-point exception. In this case, "Invalid operation". If we go to **Chapter 11.5.2.1** of volume 1, we can read:

> The floating-point invalid-operation exception occurs in response to an invalid arithmetic operand.

And from **Table 11-1**:

> Conversion to integer when the value in the source register is a NaN, ∞, or exceeds the representable range for CVTPS2PI, CVTTPS2PI, CVTSS2SI, CVTTSS2SI, CVTPD2PI, CVTSD2SI, CVTPD2DQ, CVTTPD2PI, **CVTTSD2SI**, CVTTPD2DQ, CVTPS2DQ, or CVTTPS2DQ.

As I said, this does not happen in `kernel_main`, so there is obviously something wrong when creating and switching tasks that makes the values invalid.

**Comment by 8dcc** [comment]
This problem is probably related to the fact that in `mt_newtask`, the 512 bytes that are allocated for `fxsave` and `fxrstor`, are not actually initialized.

This causes the following issue:
1. Task `A` tries to create task `B`.
2. `mt_newtask` allocates and fills the `Ctx` struct, allocates and fills the 16KiB stack with the entry point of the task, and allocates **without initializing** the 512 bytes for `Ctx.fxdata`.
3. Task `A` tries to switch to task `B`.
4. `mt_switch` stores the SSE values of the current task (`A`) in `A.fxdata` using `fxsave`.
5. `mt_switch` loads the SSE values of the new task (`B`) from `B.fxdata` using `fxrstor`

The problem is that the 512 bytes of `B.fxdata` from step 5 were not initialized in step 2.

Although right now I think I have pinned the problem, I still don't know what to use to initialize the 512 bytes of `.fxdata` when creating a new task.

**Comment by 8dcc** [comment]
Also, it's important to note that the kernel is currently enabling SSE support **without checking if it's supported first**.

From volume 3 of the Intel Developer Manual, **Chapter 13.1.2**:

> If the processor attempts to execute an unsupported SSE instruction, the processor generates an invalid-opcode exception (#UD). Before an operating system or executive attempts to use SSE extensions, it should check that support is present by confirming the following bit values returned by the CPUID instruction:
> - **CPUID.1:EDX.SSE[bit 25] = 1**
> - **CPUID.1:EDX.SSE2[bit 26] = 1**
> - CPUID.1:ECX.SSE3[bit 0] = 1
> - CPUID.1:ECX.SSSE3[bit 9] = 1
> - CPUID.1:ECX.SSE4_1[bit 19] = 1
> - CPUID.1:ECX.SSE4_2[bit 20] = 1

**Comment by 8dcc** [comment]
Calling `fxsave` with the newly allocated block inside `mt_newtask` seems to fix it, although I am not sure if this is the best way of doing it, since it means that task `B` will have the same SSE information as task `A` (the task that called `mt_newtask`).

```diff
diff --git a/src/kernel/multitask.asm b/src/kernel/multitask.asm
index f95d9de..3f06691 100644
--- a/src/kernel/multitask.asm
+++ b/src/kernel/multitask.asm
@@ -162,6 +162,7 @@ mt_newtask:
     pop     eax

     mov     [eax + ctx_t.fxdata], edx       ; Save in ctx struct
+    fxsave  [edx]

     ; Exit the mt_newtask function
     mov     esp, ebp
```

![test](https://github.com/fs-os/fs-os/assets/29655971/f1167db3-00df-46fe-afca-56d26f0a6b38)

**Comment by 8dcc** [comment]
Fixed in 9f8b884bc7bef1fd0d849cda72ce6412b3b19759, although there are going to be some minor changes.
